### ISSUE TITLE
Incomplete demon TF, when TFing from a taur-body

### ISSUE THREAD

**Issue by Stadler76** [Innoxia/liliths-throne-public#1127]
When I go through the '**Become a demon**' event starting as a taur (and probably any other non-biped leg-configuration) my legs stay unchanged. Obviously, because `LegType.DEMON_COMMON` is not allowed for demon-taurs.

In addition to that, my crotch-boobs stay unchanged, too.

A possible solution would be changing the `legType` to `LegType.DEMON_HOOFED` for taurs and change the `LegConfiguration` to `BIPEDAL` for any other non-bipeds. And to change the type of the crotch boobs, if the player has any.

Untested, but something, like this should work:
```java
switch (Main.game.getPlayer().getLegConfiguration()) {
	case BIPEDAL:
		Main.game.getPlayer().setLegType(LegType.DEMON_COMMON);
		break;

	case TAUR:
		Main.game.getPlayer().setLegType(LegType.DEMON_HOOFED);
		break;

	default:
		Main.game.getPlayer().setLegConfiguration(LegConfiguration.BIPEDAL);
		Main.game.getPlayer().setLegType(LegType.DEMON_COMMON);
}
```

**Comment by Stadler76** [comment]
Ok, I've tested it while opening this issue. Possible fix: [20190414_2147_demon-taur-fix.diff.txt](https://github.com/Innoxia/liliths-throne-public/files/3077813/20190414_2147_demon-taur-fix.diff.txt)
Note, that the comments about `setBreastCrotchSize` in that diff are hints for you, since I couldn't decide, if to set them to a fixed value or to leave them as is.

**Comment by Zsar** [comment]
Code as written does not actually require the `case BIPEDAL`.

I'd say `default` ought to throw an exception though, instead of quasi-randomly mapping to another case. This way there is actually a chance for this issue to be detected on introduction of a new leg configuration.

… Might want to then write a Unit Test for the method that iterates through `LegConfiguration.values()` - that would then be guaranteed to crash before release and can be noticed and fixed before ever going public.

The method also ought to take the character as parameter, so it works for all characters, not just the PC.

**Comment by Stadler76** [comment]
@Zsar wrote:
> Code as written does not actually require the case BIPEDAL.
> 
> I'd say default ought to throw an exception though, instead of quasi-randomly mapping to another case. This way there is actually a chance for this issue to be detected on introduction of a new leg configuration.

Agreed. But could you tell me, what exception? I'm currently doing the step from PHP -> AS3 -> Java and I haven't dug that deep into the codebase as of now.

> … Might want to then write a Unit Test for the method that iterates through LegConfiguration.values() - that would then be guaranteed to crash before release and can be noticed and fixed before ever going public.

I know how to write unit tests in AS3. But Java? No idea and there are absolutely no unit tests in the codebase to give me ideas so any help with that would be neat.

>  The method also ought to take the character as parameter, so it works for all characters, not just the PC.

I'll look into it later.

PS: Shouldn't we move this discussion to the actual PR?

**Comment by Zsar** [comment]
As to which exception would be appropriate for the default case, if the switch is kept: I'd think one of these would be appropriate:
- [IllegalArgumentException](https://docs.oracle.com/javase/8/docs/api/java/lang/IllegalArgumentException.html)
- [IllegalStateException](https://docs.oracle.com/javase/8/docs/api/java/lang/IllegalStateException.html)

(But it really does not matter much - as long as it is a RuntimeException and has a sensible message text, its type is not really interesting - as evidenced by the many `catch(Exception ex)` blocks in the code.)

**Comment by Stadler76** [comment]
This is long resolved with the merge of PR #1128. No need to keep this open anymore.

\~closing\~
